drivers/sensor/st/iis3dwb/iis3dwb.c

```diff
@@ -0,0 +1,453 @@
+	if (complete_op == NULL) {
+		return;
+	}
+
+	rtio_sqe_prep_callback_no_cqe(complete_op, complete_op_cb, (void *)dev, iodev_sqe);
+
+	rtio_submit(rtio, 0);
+}
+
+static void iis3dwb_submit_one_shot(const struct device *dev, struct rtio_iodev_sqe *iodev_sqe)
+{
+	const struct sensor_read_config *cfg = iodev_sqe->sqe.iodev->data;
+	const struct sensor_chan_spec *const channels = cfg->channels;
+	const size_t num_channels = cfg->count;
+	uint32_t min_buf_len = sizeof(struct iis3dwb_rtio_data);
+	uint64_t cycles;
+	int rc = 0;
+	uint8_t *buf;
+	uint32_t buf_len;
+	struct iis3dwb_rtio_data *edata;
+	struct iis3dwb_data *data = dev->data;
+
+	/* Get the buffer for the frame, it may be allocated dynamically by the rtio context */
+	rc = rtio_sqe_rx_buf(iodev_sqe, min_buf_len, min_buf_len, &buf, &buf_len);
+	if (rc != 0) {
+		LOG_ERR("Failed to get a read buffer of size %u bytes", min_buf_len);
+		return;
+	}
+
+	edata = (struct iis3dwb_rtio_data *)buf;
+
+	edata->has_accel = 0;
+	edata->has_temp = 0;
+
+	rc = sensor_clock_get_cycles(&cycles);
+	if (rc != 0) {
+		LOG_ERR("Failed to get sensor clock cycles");
+		rtio_iodev_sqe_err(iodev_sqe, rc);
+		return;
+	}
+
+	edata->header.is_fifo = false;
+	edata->header.range = data->range;
+	edata->header.timestamp = sensor_clock_cycles_to_ns(cycles);
+
+	for (int i = 0; i < num_channels; i++) {
+		switch (channels[i].chan_type) {
+		case SENSOR_CHAN_ACCEL_X:
+		case SENSOR_CHAN_ACCEL_Y:
+		case SENSOR_CHAN_ACCEL_Z:
+		case SENSOR_CHAN_ACCEL_XYZ:
+			edata->has_accel = 1;
+
+			uint8_t outx_regs[]  = { IIS3DWB_OUTX_L_A, };
+			struct spi_buf buf_xl[] = { {(uint8_t *)edata->accel, 6} };
+
+			/*
+			 * Prepare rtio enabled bus to read IIS3DWB_OUTX_L_A register
+			 * where accelerometer data is available.
+			 * Then iis3dwb_one_shot_complete_cb callback will be invoked.
+			 *
+			 * STMEMSC API equivalent code:
+			 *
+			 *   uint8_t accel_raw[6];
+			 *
+			 *   iis3dwb_acceleration_raw_get(&dev_ctx, accel_raw);
+			 */
+			iis3dwb_rtio_rd_transaction(dev,
+						    outx_regs,
+						    ARRAY_SIZE(outx_regs),
+						    buf_xl,
+						    iodev_sqe,
+						    iis3dwb_one_shot_complete_cb);
+			break;
+
+#if defined(CONFIG_IIS3DWB_ENABLE_TEMP)
+		case SENSOR_CHAN_DIE_TEMP:
+			edata->has_temp = 1;
+
+			uint8_t outt_regs[]  = { IIS3DWB_OUT_TEMP_L, };
+			struct spi_buf buf_t[] = { {(uint8_t *)&edata->temp, 2} };
+
+			/*
+			 * Prepare rtio enabled bus to read IIS3DWB_OUT_TEMP_L register
+			 * where temperature data is available.
+			 * Then iis3dwb_one_shot_complete_cb callback will be invoked.
+			 *
+			 * STMEMSC API equivalent code:
+			 *
+			 *   int16_t val;
+			 *
+			 *   iis3dwb_temperature_raw_get(&dev_ctx, &val);
+			 */
+			iis3dwb_rtio_rd_transaction(dev,
+						    outt_regs,
+						    ARRAY_SIZE(outt_regs),
+						    buf_t,
+						    iodev_sqe,
+						    iis3dwb_one_shot_complete_cb);
+			break;
+#endif
+
+		default:
+			continue;
+		}
+	}
+
+	if (edata->has_accel == 0) {
```

Comment:
Won't this fail on SENSOR_CHAN_DIE_TEMP only reads? Do we have to read acceleration to get the die temperature?